### PR TITLE
Define RPC types out of source

### PR DESCRIPTION
**Pull request by goldsborough**
Ran into nasty linker errors with clang-6 in fbcode. It doesn't seem to understand `static constexpr` when the value is an enum and gives me linker errors. Definining them out of source fixes the problem. Probably a compiler/linker bug

screen shot 2018-03-14 at 17 34 53

@soumith

### CODE REVIEW

File: torch/lib/THD/master_worker/common/RPCType.hpp [hunk 1]
```diff
@@ -77,79 +77,79 @@ struct type_traits {};
 
 template<>
 struct type_traits<char> {
-  static constexpr RPCType type = RPCType::CHAR;
+  static RPCType type;
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/RPCType.hpp]: It's useful to leave comments in such situations to make sure that someone won't accidentally revert this change during a refactor (because `constexpr` really makes more sense)

**Comment by ezyang** [comment]
Note that you can fix this AND preserve constexpr, see https://stackoverflow.com/questions/8016780/undefined-reference-to-static-constexpr-char

So, putting a declaration for the static constexpr in the cpp file would solve the problem. As would upgrading to C++17. And also as would, as is done here, deleting the constexpr.

If we're outlawing `static constexpr`, we should set up a lint rule to reject it.

**Comment by goldsborough** [comment]
@ezyang I'm pretty sure we weren't making use of the `constexpr`-ness of this variable since the places it was used were not `constexpr`.  AFAIS from the passages cited in the standard in that answer, this should not have compiled for us (on C++11) unless these static contexpr values were never ODR-used (their memory address never taken). From what I can see, we are actually passing them to functions that take `const&`. I think we're deep into the edge cases of the standard here where traditionally high variance can crop up within different versions of different compilers. I will follow up with better comments and the correct solution of defining the values in the cpp file.
